Wikipedia/UI-V5/WMFMainPageSectionController.m

```diff
@@ -88,24 +128,59 @@ - (void)configureCell:(UICollectionViewCell*)cell
      inCollectionView:(UICollectionView*)collectionView
           atIndexPath:(NSIndexPath*)indexPath {
     if ([cell isKindOfClass:[WMFMainPageCell class]]) {
-        WMFMainPageCell* mainPageCell = (id)cell;
+        WMFMainPageCell* mainPageCell = (WMFMainPageCell*)cell;
         mainPageCell.mainPageTitle.text = self.siteInfo.mainPageTitleText;
+    } else if ([cell isKindOfClass:[WMFArticlePreviewCell class]]) {
+        WMFArticlePreviewCell* previewCell = (WMFArticlePreviewCell*)cell;
+        previewCell.title           = [self titleForItemAtIndex:indexPath.row];
+        previewCell.descriptionText = self.featuredArticlePreview.wikidataDescription;
+        previewCell.imageURL        = self.featuredArticlePreview.thumbnailURL;
+        [previewCell setSummary:self.featuredArticlePreview.extract];
+        [previewCell setSavedPageList:self.savedPageList];
     }
 }
 
-- (void)getSiteInfo {
-    if (self.fetcher.isFetching) {
+#pragma mark - Fetching
+
+- (void)fetchData {
+    if (self.dataPromise) {
+        DDLogInfo(@"Fetch is already pending, skipping redundant call.");
         return;
     }
+
     @weakify(self);
-    [self.fetcher fetchSiteInfoForSite:self.site]
-    .then(^(id result){
+    self.dataPromise = ([self fetchFeaturedTitleIfAvailable] ? : [self fetchSiteInfo]).catch(^(NSError* error){
         @strongify(self);
-        self.siteInfo = result;
+        self.dataPromise = nil;
+        [self.delegate controller:self didFailToUpdateWithError:error];
+    });
+}
+
+- (AnyPromise*)fetchSiteInfo {
+    @weakify(self);
+    return [self.siteInfoFetcher fetchSiteInfoForSite:self.site].then(^(MWKSiteInfo* siteInfo){
+        @strongify(self);
+        self.dataPromise = nil;
+
+        self.siteInfo = siteInfo;
+        [self.delegate controller:self didSetItems:self.items];
+    });
+}
+
+- (nullable AnyPromise*)fetchFeaturedTitleIfAvailable {
```

Comment:
I'm sure I'm not able to fully understand the complications you are describing from the comments. If you want to discus options we can and figure something out. But as it is now its pretty difficult to grok - we shod be able to find something to make it more readable.